runner/runners/local_output.go

```diff
@@ -84,58 +82,123 @@ func (s *localOutputCreator) Create(id string) (runner.Output, error) {
-        }
-        var wasAtBottom = checkAtBottom()
-        var div = document.getElementById("output");
-        var content = document.createTextNode(txt);
-        div.appendChild(content);
-        if (wasAtBottom)
-          gotoBottom()
-        if (id != resourceId) {
-          throw new Error('ResourceId mismatch.')
-        }
-      }
-    }
-    //TODO: request range to get delta.
-    xhr.open("GET", location.href + (location.search=="" ? "?" : "&") + "content=true");
-    xhr.send();
-  };
-  sendRequest()
-  setInterval(sendRequest, 5000)
-</script>
-<body><div id="output" style="white-space: pre-wrap"></div></body>
-</html>
+		<script type="text/javascript">
+			let resourceId = "";
+			const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
+			const checkAtBottom = () => {
+				//scrolling: http://stackoverflow.com/a/22394544
+				let scrollTop =
+					(document.documentElement && document.documentElement.scrollTop) ||
+					document.body.scrollTop;
+				let scrollHeight =
+					(document.documentElement && document.documentElement.scrollHeight) ||
+					document.body.scrollHeight;
+				return scrollTop + window.innerHeight >= scrollHeight;
+			};
+			const gotoBottom = () => {
+				let scrollHeight =
+					(document.documentElement && document.documentElement.scrollHeight) ||
+					document.body.scrollHeight;
+				let scrollLeft =
+					(document.documentElement && document.documentElement.scrollLeft) ||
+					document.body.scrollLeft;
+				window.scrollTo(scrollLeft, scrollHeight);
+			};
+			const getLength = resp => {
+				let contentRange = resp.headers.get("Content-Range");
+				let idx = contentRange.lastIndexOf("/") + 1;
+				return contentRange.slice(idx);
+			};
+	
+			const copyBuffer = (oldBuffer, length) => {
+				buffer = new Uint8Array(length);
+				buffer.set(oldBuffer);
+				return {
+					buffer,
+					length
+				};
+			};
+	
+			const writeBuffer = (mainBuffer, arrayBuffer, offset) => {
+				let bytes = new Uint8Array(arrayBuffer);
+				for (let i = 0; i < bytes.byteLength; i++) {
+					mainBuffer[i + offset] = bytes[i];
+				}
+			};
+	
+			const updateText = buffer => {
+				let wasAtBottom = checkAtBottom();
+				let div = document.getElementById("output");
+				div.innerText = new TextDecoder("utf-8").decode(buffer);
+				if (wasAtBottom) {
+					gotoBottom();
+				}
+			};
+	
+			const increaseTimeout = currTimeout => Math.trunc((currTimeout * 3) / 2);
+	
+			sendRequest = async () => {
+				let url =
+					location.href + (location.search == "" ? "?" : "&") + "content=true";
+				let resp = await fetch(url, {
+					method: "HEAD"
+				});
+				let length = Number.parseInt(resp.headers.get("Content-Length"));
+				let buffer = new Uint8Array(length);
+				let curr = 0;
+				// 128KB == 131072
+				let offset = 1310720;
+				let minTimeout = 50;
+				let currTimeout = minTimeout;
+				// 15 minutes
+				let maxTimeout = 15 * 60 * 1000;
+				while (true) {
```

Comment:
can we have a comment explaining the overall approach... ?getting the log in batches of 128KB bytes?

Reply:
+1

Reply:
Done